### PR TITLE
ci: Split out native fuzz jobs for macOS and windows (take 2)

### PR DESCRIPTION
**Pull request by dergoegge**
Split out two new CI jobs (for native macOS and windows) that run the fuzz tests on the qa-assets input corpora.

In both jobs the fuzz binary is built with `-DBUILD_FOR_FUZZING` to enable `Assume` assertions as well as `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION`.

### CODE REVIEW

**Comment by dergoegge** [comment]
Looks like using the matrix feature works for this

**Comment by dergoegge** [comment]
Now it works, ready for review!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -74,20 +74,27 @@ jobs:
           # Use clang++, because it is a bit faster and uses less memory than g++
           git rebase --exec "echo Running test-one-commit on \$( git log -1 ) && CC=clang CXX=clang++ cmake -B build -DWERROR=ON -DWITH_ZMQ=ON -DBUILD_GUI=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DWITH_BDB=ON -DWITH_USDT=ON -DCMAKE_CXX_FLAGS='-Wno-error=unused-member-function' && cmake --build build -j $(nproc) && ctest --output-on-failure --test-dir build -j $(nproc) && ./build/test/functional/test_runner.py -j $(( $(nproc) * 2 ))" ${{ env.TEST_BASE }}
 
-  macos-native-arm64:
-    name: 'macOS 14 native, arm64, no depends, sqlite only, gui'
-    # Use latest image, but hardcode version to avoid silent upgrades (and breaks).
-    # See: https://github.com/actions/runner-images#available-images.
+  macos-native:
+    name: 'macOS 14 native'
```
Comment (hebasto) [inline .github/workflows/ci.yml:78]: You might want to use `${{ matrix.conf.job_name }}`.

Same for the Windows job.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -178,30 +193,26 @@ jobs:
...
-          cmake -B build --preset vs2022-static -DCMAKE_TOOLCHAIN_FILE="$env:VCPKG_INSTALLATION_ROOT\scripts\buildsystems\vcpkg.cmake" -DBUILD_GUI=ON -DWITH_BDB=ON -DWITH_ZMQ=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DWERROR=ON
+          cmake -B build --preset vs2022-static -DCMAKE_TOOLCHAIN_FILE="$env:VCPKG_INSTALLATION_ROOT\scripts\buildsystems\vcpkg.cmake" ${{ matrix.generate-options }}
 
       - name: Save vcpkg binary cache
         uses: actions/cache/save@v4
         if: github.event_name != 'pull_request' && steps.vcpkg-binary-cache.outputs.cache-hit != 'true'
         with:
           path: ~/AppData/Local/vcpkg/archives
-          key: ${{ github.job }}-vcpkg-binary-${{ hashFiles('cmake_version', 'msbuild_version', 'toolset_version', 'vcpkg.json') }}
+          key: ${{ matrix.job-type }}-vcpkg-binary-${{ hashFiles('cmake_version', 'msbuild_version', 'toolset_version', 'vcpkg.json') }}
```
Comment (hebasto) [inline .github/workflows/ci.yml:207]: Instead of saving another cache, we can simply skip the cache saving step for the "fuzz" job type.

Comment (dergoegge) [reply .github/workflows/ci.yml]: Got rid of all the duplicate caches

File: .github/workflows/ci.yml (hunk 1 above)
Comment (dergoegge) [reply .github/workflows/ci.yml]: Doesn't look like this works

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -142,6 +154,17 @@ jobs:
...
+    strategy:
+      matrix:
+        job-type: [standard, fuzz]
+        include:
+          - job-type: standard
+            generate-options: '-DBUILD_GUI=ON -DWITH_BDB=ON -DWITH_ZMQ=ON -DBUILD_BENCH=ON -DWERROR=ON'
+            job-name: 'Win64 native, VS 2022'
+          - job-type: fuzz
+            generate-options: '-DBUILD_FOR_FUZZING=ON -DWERROR=ON'
+            job-name: 'Win64 native fuzz, VS 2022'
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:166]: > It seems a bit annoying that the Windows build is now duplicated and seems to take 2h each times 2 = 4 CI hours total.

I assumed this is a caching issue? This shouldn't be doing anything different from the current windows CI...

[the remark this answers is not in the document]
Comment (fanquake) [reply .github/workflows/ci.yml:166]: The majority of the time here is vcpkg compiling Qt modules from source, in the "Generate build system step": https://github.com/bitcoin/bitcoin/actions/runs/11705992609/job/32602049089?pr=31221. That should be getting cached in some way. Maybe push again and see if things are rebuilt?

[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:166]: Re-pushed

[the remark this answers is not in the document]
Comment (fanquake) [reply .github/workflows/ci.yml:166]: I would also think we can just turn off everything GUI related off for fuzzing?

[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:166]: > Looks like it's rebuilding everything again, could it be that it can't store new caches on PRs?

The pkg cache is not stored on PRs (see [here](https://github.com/bitcoin/bitcoin/blob/2a0997b3a5dae13e1b42776123bc9a69781d2222/.github/workflows/ci.yml#L212)).

But I think the main problem was that I changed the job name from `win64-native` to `windows-native` which changed the lookup key for the cache (I assume).  Re-pushed again.

[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:166]: "Cache restored successfully" 🎉. "Generate build system" only took 4min now. Resolving.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -120,6 +130,8 @@ jobs:
 
       - name: CI script
         run: ./ci/test_run_all.sh
+        env:
+          FILE_ENV: ${{ matrix.file-env }}
 
       - name: Save Ccache cache
         uses: actions/cache/save@v4
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:137]: Hm, should we have a separate ccache for each job type? I assume, there is no way to merge the ccache from the two

[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:137]: > have a separate ccache for each job type

Done

**Comment by dergoegge** [comment]
@hebasto could you give this another look?

**Comment by fanquake** [comment]
Concept ACK.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -85,9 +85,19 @@ jobs:
...
+    strategy:
+      matrix:
+        job-type: [standard, fuzz]
+        include:
+          - job-type: standard
+            file-env: './ci/test/00_setup_env_mac_native.sh'
+            job-name: 'macOS 14 native, arm64, no depends, sqlite only, gui'
+          - job-type: fuzz
+            file-env: './ci/test/00_setup_env_mac_native_fuzz.sh'
+            job-name: 'macOS 14 native arm64 fuzz'
```
Comment (hebasto) [inline .github/workflows/ci.yml:97]: Might be separated with commas for consistency with the other `job-name`?

File: ci/test/00_setup_env_mac_native_fuzz.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Copyright (c) The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+export LC_ALL=C.UTF-8
+
+# Homebrew's python@3.12 is marked as externally managed (PEP 668).
+# Therefore, `--break-system-packages` is needed.
+export PIP_PACKAGES="--break-system-packages zmq"
```
Comment (hebasto) [inline ci/test/00_setup_env_mac_native_fuzz.sh:11]: Why are these lines needed?

**Review by hebasto — COMMENTED** [review]
Approach ACK c3354ea6dc34a3cadc60346c363478af22a5c0c2.

Please add [`strategy.fail-fast: false`](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast) everywhere.

File: ci/test/00_setup_env_mac_native_fuzz.sh (hunk 6 above)
Comment (dergoegge) [reply ci/test/00_setup_env_mac_native_fuzz.sh]: They're probably not, removed

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -142,6 +154,17 @@ jobs:
...
 
+    strategy:
+      matrix:
+        job-type: [standard, fuzz]
+        include:
+          - job-type: standard
+            generate-options: '-DBUILD_GUI=ON -DWITH_BDB=ON -DWITH_ZMQ=ON -DBUILD_BENCH=ON -DWERROR=ON'
+            job-name: 'Win64 native, VS 2022'
+          - job-type: fuzz
+            generate-options: '-DBUILD_FOR_FUZZING=ON -DWERROR=ON'
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml:167]: `-DBUILD_GUI=OFF` is implied by `-DBUILD_FOR_FUZZING=ON`.

What does `-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="sqlite"` do? we're using the same vcpkg cache for both jobs would that still work with this change?

[the remark this answers is not in the document]
Comment (hebasto) [reply .github/workflows/ci.yml:167]: > `-DBUILD_GUI=OFF` is implied by `-DBUILD_FOR_FUZZING=ON`.

Yes. But it is evaluated _after_ an attempt to look for the Qt packages. In this case, `-DBUILD_GUI=OFF`  simply negates `-DBUILD_GUI=ON` from the "vs2022-static" preset.

> What does `-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="sqlite"` do? 

The optional vcpkg packages are organised in "features" in the [`vcpkg.json`](https://github.com/bitcoin/bitcoin/blob/master/vcpkg.json) manifest file.

`-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON -DVCPKG_MANIFEST_FEATURES="sqlite"` means "build and install only the `sqlite3` package, skipping all other optional packages".

> we're using the same vcpkg cache for both jobs would that still work with this change?

To be precise, both jobs read the vcpkg binary cache, which works for a subset of packages just fine, and only the "standard" job writes the cache, which is correct.

[the remark this answers is not in the document]
Comment (dergoegge) [reply .github/workflows/ci.yml]: Done.

File: ci/test/00_setup_env_mac_native_fuzz.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) The Bitcoin Core developers
```
Comment (hebasto) [inline ci/test/00_setup_env_mac_native_fuzz.sh:3]: nit: Add a year?
```suggestion
# Copyright (c) 2024-present The Bitcoin Core developers
```

**Review by hebasto — APPROVED** [review]
ACK b031b7910d62819443813b91705211c466933a53.

File: ci/test/00_setup_env_mac_native_fuzz.sh (hunk 8 above)
Comment (dergoegge) [reply ci/test/00_setup_env_mac_native_fuzz.sh:3]: I'll leave this as is unless another reviewer prefers having the year as well.

**Comment by dergoegge** [comment]
rfm?

**Comment by dergoegge** [comment]
@achow101 @ryanofsky @glozow Would appreciate if one of you could take a look :)
